core/client/client.go

```diff
@@ -34,9 +38,48 @@ var (
 	errNoPaths = errors.New("failed to measure clock offset: no paths")
 )
 
+func printmeta(meta ntske.Data, log *zap.Logger) {
+	log.Debug("NTSKE exchange yielded:",
+		zap.String("c2s", hex.EncodeToString(meta.C2sKey)),
+		zap.String("s2c", hex.EncodeToString(meta.S2cKey)),
+		zap.String("server", meta.Server),
+		zap.Uint16("port", meta.Port),
+		zap.Uint16("algo", meta.Algo))
+
+	log.Debug("Cookies: ", zap.Int("number of ", len(meta.Cookie)))
+
+	for i, cookie := range meta.Cookie {
+		log.Debug(strconv.Itoa(i), zap.String(".", hex.EncodeToString(cookie)))
+	}
+}
+
 func MeasureClockOffsetIP(ctx context.Context, log *zap.Logger,
 	ntpc *IPClient, localAddr, remoteAddr *net.UDPAddr) (
 	time.Duration, error) {
+
+	if ntpc.IsAuthorized {
+		// First do key exchange and save Keys and cookies in ntpc struct
+		// - TCP TLS handshake
+		// - KE
+		// - Must store port and server somewhere
+		//    - could use the ntske Data struct and store in IPNTSClient
+		if ntpc.auth.keyExchangeNTS == nil {
+			tlsconfig := &tls.Config{}
+
+			// For testing do not verify tls certificate
+			// CHANGE IN PRODUCTION VERSION
+			tlsconfig.InsecureSkipVerify = true
+			server := remoteAddr.IP.String() //For TLS certificate we need the string dns address. Otherwise use remoteAddr.IP.String()
```

Comment:
Simplified code to now use the tls.Config.ServerName in the following commit
[use server from tls.Config for keyExchange](https://github.com/marcfrei/scion-time/pull/2/commits/db53f06edb81dc094d55c57b30c64286547e0b15)

It is still unclear to me though how to best set the ServerName in the config. Can you retrieve the domain name from the net.UdpAddr struct? Should we pass it seperately or do we even need it at all for SCION?

Reply:
In my opinion, It would be ideal if the entire IPClient.Auth including the sub-struct TLSConfig could be set up in the main package based on the provided command-line arguments (and later on maybe also based on a config file). Does this make sense?

Reply:
Yes makes sense thank you.